### PR TITLE
Advance the state of ngtsc's template type-checking to the point where it can handle AIO.

### PR DESCRIPTION
**Pull request by alxhub**

### CODE REVIEW

**Review by JoostK — COMMENTED** [review]
Truly awesome work! 🎉

File: packages/compiler-cli/src/ngtsc/typecheck/src/expression.ts [hunk 1]
```diff
@@ -76,6 +76,22 @@ export function astToTypescript(
     const trueExpr = astToTypescript(ast.trueExp, maybeResolve);
     const falseExpr = astToTypescript(ast.falseExp, maybeResolve);
     return ts.createParen(ts.createConditional(condExpr, trueExpr, falseExpr));
+  } else if (ast instanceof LiteralArray) {
+    const elements = ast.expressions.map(expr => astToTypescript(expr, maybeResolve));
+    return ts.createArrayLiteral(elements);
+  } else if (ast instanceof LiteralMap) {
```
[the remark this answers is not in the document]
Comment (alxhub) [reply packages/compiler-cli/src/ngtsc/typecheck/src/expression.ts]: Yes, I believe so!

File: packages/compiler-cli/src/ngtsc/typecheck/src/type_check_block.ts [hunk 2]
```diff
@@ -46,9 +41,267 @@ export function generateTypeCheckBlock(
...
+    return null;
+  }
+}
+
+/**
+ * Value used to break a circular reference between `TcbOp`s.
+ *
+ * This value is returned whenever `TcbOp`s have a circular dependency. The expression is a non-null
+ * assertion of the null value (in TypeScript, the expression `null!`). This construction will infer
+ * the bottom type for whatever it's assigned to.
```
[the remark this answers is not in the document]
Comment (alxhub) [reply packages/compiler-cli/src/ngtsc/typecheck/src/type_check_block.ts]: Hm, I could be more precise here.

File: packages/compiler-cli/src/ngtsc/typecheck/src/type_check_block.ts [hunk 3]
```diff
@@ -610,9 +619,11 @@ class Scope {
     } else if (node instanceof TmplAstTemplate) {
       // Template children are rendered in a child scope.
       this.appendDirectivesAndInputsOfNode(node);
-      const ctxIndex = this.opQueue.push(new TcbTemplateContextOp(this.tcb, this, node)) - 1;
-      this.templateCtxOpMap.set(node, ctxIndex);
-      this.opQueue.push(new TcbTemplateBodyOp(this.tcb, this, node));
+      if (this.tcb.config.checkTemplateBodies) {
```
[the remark this answers is not in the document]
Comment (alxhub) [reply packages/compiler-cli/src/ngtsc/typecheck/src/type_check_block.ts]: Believe it or not, it really was that limited.

File: packages/compiler-cli/src/ngtsc/metadata/src/inheritance.ts [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+    outputs = {...outputs, ...meta.outputs};
+  };
+
+  addMetadata(topMeta);
+
+  return {
+    ...topMeta,
+    inputs,
+    outputs,
+    baseClass: isDynamic ? 'dynamic' : null,
```
[the remark this answers is not in the document]
Comment (alxhub) [reply packages/compiler-cli/src/ngtsc/metadata/src/inheritance.ts]: It doesn't affect it right now, correct. I need to think more on how this should work, but I wanted to make sure to distinguish the two cases for later.

File: packages/compiler-cli/src/ngtsc/typecheck/src/expression.ts [hunk 5]
```diff
@@ -6,10 +6,15 @@
  * found in the LICENSE file at https://angular.io/license
  */
 
-import {AST, ASTWithSource, Binary, Conditional, Interpolation, KeyedRead, LiteralArray, LiteralMap, LiteralPrimitive, MethodCall, NonNullAssert, PropertyRead} from '@angular/compiler';
+import {AST, ASTWithSource, Binary, Conditional, Interpolation, KeyedRead, LiteralArray, LiteralMap, LiteralPrimitive, MethodCall, NonNullAssert, PrefixNot, PropertyRead, SafeMethodCall, SafePropertyRead} from '@angular/compiler';
 import * as ts from 'typescript';
 
 import {TypeCheckingConfig} from './api';
+
+const NULL_AS_ANY =
```
[the remark this answers is not in the document]
Comment (alxhub) [reply packages/compiler-cli/src/ngtsc/typecheck/src/expression.ts]: Yes, because we're not putting it through the binding operation (which would happen if we added it to ` ts.SourceFile` as part of a `ts.CompilerHost` operation.

File: packages/compiler-cli/src/ngtsc/typecheck/src/type_check_file.ts [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+import {NoopImportRewriter, Reference, ReferenceEmitter} from '../../imports';
+import {AbsoluteFsPath} from '../../path';
+import {ClassDeclaration} from '../../reflection';
+import {ImportManager} from '../../translator';
+
+import {TypeCheckBlockMetadata, TypeCheckingConfig} from './api';
+import {Environment} from './environment';
+import {generateTypeCheckBlock} from './type_check_block';
+
+export class TypeCheckFile extends Environment {
```
[the remark this answers is not in the document]
Comment (alxhub) [reply packages/compiler-cli/src/ngtsc/typecheck/src/type_check_file.ts]: From a practical perspective, it uses some of the protected fields.

From a theoretical standpoint, a `TypeCheckFile` _is an_ `Environment` - it's a context in which type-check block(s) are hosted. The `Environment` class can function standalone (e.g. for tests, or when generating a single TCB inline). The `TypeCheckFile` is a whole file environment in which multiple TCBs can be rendered.

[the remark this answers is not in the document]
Comment (JoostK) [reply packages/compiler-cli/src/ngtsc/typecheck/src/type_check_file.ts]: Ah, thanks that makes sense. `Environment` is such a generic name that I didn't consider it as such. Maybe `TcbEnvironment` would make it clearer.

File: packages/compiler-cli/src/ngtsc/typecheck/test/type_check_block_spec.ts [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+
+class FakeRefEmitter /* implements ReferenceEmitter */ {
+  emit(ref: Reference<ts.Node>, context: ts.SourceFile, importMode?: ImportMode): Expression {
+    return new ExternalExpr({
+      moduleName: `types/${ref.debugName}`,
+      name: ref.debugName,
+    });
+  }
+
+  static newFake(): ReferenceEmitter { return new FakeRefEmitter() as ReferenceEmitter; }
```
[the remark this answers is not in the document]
Comment (alxhub) [reply packages/compiler-cli/src/ngtsc/typecheck/test/type_check_block_spec.ts]: 👍 good idea. Done.

File: packages/compiler-cli/src/ngtsc/typecheck/src/type_check_block.ts [hunk 8]
```diff
@@ -95,160 +384,261 @@ class Context {
...
+        (ref instanceof TmplAstElement || ref instanceof TmplAstTemplate) &&
+        directive !== undefined && this.directiveOpMap.has(ref)) {
+      // Resolving a directive on an element or sub-template.
+      const dirMap = this.directiveOpMap.get(ref) !;
+      if (dirMap.has(directive)) {
+        return this.resolveOp(dirMap.get(directive) !);
+      } else {
+        return null;
+      }
+    } else if (ref instanceof TmplAstElement && this.elementOpMap.has(ref as TmplAstElement)) {
```
[the remark this answers is not in the document]
Comment (alxhub) [reply packages/compiler-cli/src/ngtsc/typecheck/src/type_check_block.ts]: Done.

File: packages/compiler-cli/src/ngtsc/typecheck/src/type_check_block.ts [hunk 9]
```diff
@@ -86,6 +339,42 @@ class Context {
...
+   *
+   * This may involve importing the node into the file if it's not declared there already.
+   */
+  referenceType(ref: Reference<ts.Node>): ts.TypeNode {
+    const ngExpr = this.refEmitter.emit(ref, this.sourceFile);
+    if (ngExpr === null) {
+      throw new Error(`Unreachable reference: ${ref.node}`);
+    }
+
+    // Create an `ExpressionType` from the `Expression` and translate it via `translateType`.
```
[the remark this answers is not in the document]
Comment (alxhub) [reply packages/compiler-cli/src/ngtsc/typecheck/src/type_check_block.ts]: I don't mind it; to me it's actually about the explicitness. This makes it clear that we're intending to use `translateType` by creating a synthetic `ExpressionType` node. This does answer "why", as opposed to just reading the code where it may not be obvious that the `ExpressionType` is intended to be synthetic.

File: packages/compiler-cli/src/ngtsc/typecheck/src/type_check_block.ts [hunk 10]
```diff
@@ -46,9 +41,267 @@ export function generateTypeCheckBlock(
...
+}
+
+/**
+ * Value used to break a circular reference between `TcbOp`s.
+ *
+ * This value is returned whenever `TcbOp`s have a circular dependency. The expression is a non-null
+ * assertion of the null value (in TypeScript, the expression `null!`). This construction will infer
+ * the bottom type for whatever it's assigned to.
+ */
+const INFER_TYPE_FOR_CIRCULAR_OP_EXPR = ts.createNonNullExpression(ts.createNull());
```
[the remark this answers is not in the document]
Comment (alxhub) [reply packages/compiler-cli/src/ngtsc/typecheck/src/type_check_block.ts]: Yes, it's fine to reuse nodes.

Re: the test - agreed, added to `type_check_block_spec.ts`.

File: packages/compiler-cli/src/ngtsc/program.ts [hunk 11]
```diff
@@ -378,15 +375,47 @@ export class NgtscProgram implements api.Program {
...
+    }
+
+    const compilation = this.ensureAnalyzed();
+
+    // Run template type-checking.
+
+    // First select a type-checking configuration, based on whether full template type-checking is
+    // requested.
+    let typeCheckingConfig: TypeCheckingConfig;
+    if (this.options.fullTemplateTypeCheck) {
```
[the remark this answers is not in the document]
Comment (alxhub) [reply packages/compiler-cli/src/ngtsc/program.ts]: 👍

File: packages/compiler-cli/src/ngtsc/typecheck/src/type_check_file.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (alxhub) [reply packages/compiler-cli/src/ngtsc/typecheck/src/type_check_file.ts]: I added comments to both `Environment` and `TypeCheckFile` explaining their purpose and relationship.

File: packages/bazel/src/ng_module.bzl [hunk 12]
```diff
@@ -299,6 +299,9 @@ def _ngc_tsconfig(ctx, files, srcs, **kwargs):
         "enableSummariesForJit": is_legacy_ngc,
         "enableIvy": _enable_ivy_value(ctx),
         "fullTemplateTypeCheck": ctx.attr.type_check,
+        # TODO(alxhub/arick): template type-checking for Ivy needs to be tested in g3 before it can
+        # be enabled here.
+        "ivyTemplateTypeCheck": False,
```
Comment (alexeagle) [inline packages/bazel/src/ng_module.bzl]: how is this related to `_rollout_template_typecheck` already in g3?

Comment (alxhub) [reply packages/bazel/src/ng_module.bzl]: `_rollout_template_typecheck` is internal to the `ng_module` rule and controls the `fullTemplateTypeCheck` flag. We haven't vetted Ivy's type-checking implementation against g3 at all.

This PR makes `ngtsc` perform Ivy's checks by default. This _should_ be as lenient as VE's implementation, but there might be corner cases. Therefore, to avoid blocking the ongoing runtime work in g3 by breaking builds with type-checking errors, I'm temporarily disabling type-checking in ngtsc in g3 until we can pass a global presubmit with this flag enabled. Then we can delete the flag.

File: packages/compiler-cli/src/ngtsc/typecheck/test/type_check_block_spec.ts [hunk 13]
```diff
@@ -39,6 +39,42 @@ describe('type check blocks', () => {
...
+  const TEMPLATE = `
+  <div dir #d="dir" [input]="d"></div>
+`;
+  const DIRECTIVES: TestDirective[] = [{
+    name: 'Dir',
+    selector: '[dir]',
+    exportAs: ['dir'],
+    inputs: {input: 'input'},
+  }];
+  expect(tcb(TEMPLATE, DIRECTIVES)).toContain('var _t2 = i0.Dir.ngTypeCtor({ input: (null!) });');
```
Comment (petebacondarwin) [inline packages/compiler-cli/src/ngtsc/typecheck/test/type_check_block_spec.ts]: How does this work? It seems that we are checking that the type of the value passed to the input attribute is defined? Why is it not passing `_t2` rather than `null!`?

File: packages/compiler-cli/src/ngtsc/typecheck/src/type_check_block.ts [hunk 14]
```diff
@@ -46,9 +41,267 @@ export function generateTypeCheckBlock(
       /* typeParameters */ node.typeParameters,
       /* parameters */[tcbCtxParam(node)],
       /* type */ undefined,
-      /* body */ body);
+      /* body */ ts.createBlock([ts.createIf(ts.createTrue(), scope.renderToBlock())]));
+}
+
+/**
+ * A code generation ooperation that's part of constructing a Type Check Block.
```
Comment (petebacondarwin) [inline packages/compiler-cli/src/ngtsc/typecheck/src/type_check_block.ts]: ooperation => operation

and I don't really understand what `that's part of constructing` means. Is it equivalent to `that's used to construct`?

File: packages/compiler-cli/src/ngtsc/typecheck/src/type_check_block.ts [hunk 15]
```diff
@@ -46,9 +41,267 @@ export function generateTypeCheckBlock(
...
+ * directive, extracting a variable from a let- operation, etc), then it's possible for operations
+ * earlier in the sequence to depend on operations which occur later in the sequence.
+ *
+ * `TcbOp` abstracts the different types of operations which are required to convert a template into
+ * a TCB. This allows for two phases of processing for the template, where 1) a linear sequence of
+ * `TcbOp`s is generated, and then 2) these operations are executed, not necessarily in linear
+ * order.
+ *
+ * Each `TcbOp` may insert statements into the body of the TCB, and also optionally return a
+ * `ts.Expression` which can be used to reference the operation's result.
```
Comment (petebacondarwin) [inline packages/compiler-cli/src/ngtsc/typecheck/src/type_check_block.ts]: So effectively what we are talking about is "hoisting" template variables, right? Is the scope of the hoisting the whole template or is it more fine-grained (e.g. a containing element)?

File: packages/compiler-cli/src/ngtsc/typecheck/src/type_check_block.ts [hunk 16]
```diff
@@ -46,9 +41,267 @@ export function generateTypeCheckBlock(
...
+
+/**
+ * A `TcbOp` which descends into a `TmplAstTemplate`'s children and generates type-checking code for
+ * them.
+ *
+ * This operation wraps the children's type-checking code in an `if` block, which may include one
+ * or more type guard conditions that narrow types within the template body.
+ */
+class TcbTemplateBodyOp extends TcbOp {
+  constructor(private tcb: Context, private scope: Scope, private node: TmplAstTemplate) {
```
Comment (petebacondarwin) [inline packages/compiler-cli/src/ngtsc/typecheck/src/type_check_block.ts]: I would name `node` => `template` or `templateBody`.

Comment (petebacondarwin) [reply packages/compiler-cli/src/ngtsc/typecheck/src/type_check_block.ts]: And similarly for other ops, `node` is very generic and actually decreases readability of the later code in each op class.

File: packages/compiler-cli/src/ngtsc/typecheck/test/type_check_block_spec.ts (hunk 13 above)
Comment (petebacondarwin) [reply packages/compiler-cli/src/ngtsc/typecheck/test/type_check_block_spec.ts]: How about a test where there are two directives that refer to each other?

File: packages/compiler-cli/src/ngtsc/typecheck/src/type_check_block.ts [hunk 17]
```diff
@@ -633,9 +809,27 @@ function tcbResolve(ast: AST, tcb: Context, scope: Scope): ts.Expression|null {
...
+        }
+
+        if (target instanceof TmplAstElement) {
+          return scope.resolve(target);
+        } else if (target instanceof TmplAstTemplate) {
+          // For direct template references.
+          let value: ts.Expression = ts.createNull();
+          value = ts.createAsExpression(value, ts.createKeywordTypeNode(ts.SyntaxKind.AnyKeyword));
+          value = ts.createAsExpression(value, tcb.referenceCoreType('TemplateRef', 1));
+          value = ts.createParen(value);
```
Comment (petebacondarwin) [inline packages/compiler-cli/src/ngtsc/typecheck/src/type_check_block.ts]: So this generates `(null as any as TemplateRef)`, yes?

File: packages/compiler-cli/ngcc/src/analysis/decoration_analyzer.ts [hunk 18]
```diff
@@ -6,6 +6,7 @@
  * found in the LICENSE file at https://angular.io/license
  */
 import {ConstantPool} from '@angular/compiler';
+import {DtsMetadataReader, LocalMetadataRegistry} from '@angular/compiler-cli/src/ngtsc/metadata';
```
Comment (petebacondarwin) [inline packages/compiler-cli/ngcc/src/analysis/decoration_analyzer.ts]: Although VS Code auto-imports using `@angular/compiler-cli/src/ngtsc/...` this is a deep import and I believe that we are more inclined to use a relative import in these cases, e.g. `../../../src/ngtsc/metdata`.

File: packages/compiler-cli/src/ngtsc/metadata/src/api.ts [hunk 19]
```diff
@@ -0,0 +1,70 @@
...
+/**
+ * Metadata for a pipe within an NgModule's scope.
+ */
+export interface PipeMeta {
+  ref: Reference<ClassDeclaration>;
+  name: string;
+}
+
+/**
+ * Reads metadata for directives, pipes, and modules.
```
Comment (petebacondarwin) [inline packages/compiler-cli/src/ngtsc/metadata/src/api.ts]: OOC reads from where?

File: packages/compiler-cli/src/ngtsc/metadata/src/util.ts [hunk 20]
```diff
@@ -93,3 +95,36 @@ function nodeStaticMethodNames(node: ClassDeclaration, reflector: ReflectionHost
       .filter(member => member.kind === ClassMemberKind.Method && member.isStatic)
       .map(member => member.name);
 }
+
+export class CompoundMetadataReader implements MetadataReader {
```
Comment (petebacondarwin) [inline packages/compiler-cli/src/ngtsc/metadata/src/util.ts]: A brief doc-comment would be helpful here, especially to explain why this returns only the first non-null result from the readers.

File: packages/compiler-cli/test/ngtsc/template_typecheck_spec.ts [hunk 21]
```diff
@@ -164,4 +164,46 @@ describe('ngtsc type checking', () => {
     expect(diags.length).toBe(1);
     expect(diags[0].messageText).toContain('does_not_exist');
   });
+
+  it('should properly type-check inherited directives', () => {
+    env.write('test.ts', `
+    import {Component, Directive, Input, NgModule} from '@angular/core';
+
+    @Directive({
+      selector: '[dir]',
```
Comment (petebacondarwin) [inline packages/compiler-cli/test/ngtsc/template_typecheck_spec.ts]: Although this directive will not match directly because it is not listed in the `declarations` of the `NgModule`, it is a bit confusing for both directives (`BaseDir` and `ChildDir`) to match the `[dir]` attribute. If both were listed in the `NgModule` then this test would fail even if we were not looking up the inheritance stack.

Is that correct? If so, I would change the selector on the `BaseDir` so that it could never match in the test template.

File: packages/compiler-cli/src/ngtsc/typecheck/src/expression.ts [hunk 22]
```diff
@@ -94,6 +99,20 @@ export function astToTypescript(
   } else if (ast instanceof NonNullAssert) {
     const expr = astToTypescript(ast.expression, maybeResolve, config);
     return ts.createNonNullExpression(expr);
+  } else if (ast instanceof PrefixNot) {
+    return ts.createLogicalNot(astToTypescript(ast.expression, maybeResolve, config));
+  } else if (ast instanceof SafePropertyRead) {
+    const receiver = astToTypescript(ast.receiver, maybeResolve, config);
+    const expr = ts.createPropertyAccess(ts.createNonNullExpression(receiver), ast.name);
```
Comment (petebacondarwin) [inline packages/compiler-cli/src/ngtsc/typecheck/src/expression.ts]: If the `safeTernary()` has checked that `receiver != null` then why to we need to add the non-null operator to it?

i.e. isn't `a != null ? a!.b : null` the same as `a != null ? a.b : null`?

**Comment by petebacondarwin** [comment]
The commit message

> perf(ivy): template type-check the entire program in 1 file if possible

has a typo:

> parameters are present, to the approach taken is actually a hybrid. These

-> so the approach

**Review by petebacondarwin — APPROVED** [review]
Amazing work @alxhub - quite an opus. Looks good - a few minor nits and it needs rebase conflicts resolving.

File: packages/compiler-cli/src/ngtsc/typecheck/test/type_check_block_spec.ts (hunk 13 above)
Comment (alxhub) [reply packages/compiler-cli/src/ngtsc/typecheck/test/type_check_block_spec.ts]: You can't use a variable in the initializer of its own declaration. Passing `_t2` would be equivalent to `var x = x;`

File: packages/compiler-cli/src/ngtsc/typecheck/src/type_check_block.ts (hunk 17 above)
Comment (alxhub) [reply packages/compiler-cli/src/ngtsc/typecheck/src/type_check_block.ts]: Correct :)

File: packages/compiler-cli/src/ngtsc/typecheck/src/type_check_block.ts (hunk 14 above)
Comment (alxhub) [reply packages/compiler-cli/src/ngtsc/typecheck/src/type_check_block.ts]: > A code generation operation that's involved in the construction of a Type Check Block.

File: packages/compiler-cli/src/ngtsc/typecheck/src/type_check_block.ts (hunk 15 above)
Comment (alxhub) [reply packages/compiler-cli/src/ngtsc/typecheck/src/type_check_block.ts]: Yes, exactly. We're hoisting declarations of things in the template when we need to refer to them sooner than they would otherwise be declared if the template were processed in the traditional depth-first order.

It's fine-grained, so a declaration will be moved immediately before the point where it's first used, if it's needed before its natural position.

File: packages/compiler-cli/src/ngtsc/typecheck/src/type_check_block.ts [hunk 23]
```diff
@@ -79,13 +332,43 @@ class Context {
...
-    if (ngExpr === null) {
-      throw new Error(`Unreachable reference: ${ref.node}`);
-    }
 
     // Use `translateExpression` to convert the `Expression` into a `ts.Expression`.
     return translateExpression(ngExpr, this.importManager, NOOP_DEFAULT_IMPORT_RECORDER);
   }
+
+  /**
+   * Write a `ts.TypeNode` that references the given node as a type.
```
[the remark this answers is not in the document]
Comment (alxhub) [reply packages/compiler-cli/src/ngtsc/typecheck/src/type_check_block.ts]: You're absolutely right - changed to "Generate" instead of "Write".

File: packages/compiler-cli/src/ngtsc/typecheck/src/type_check_block.ts (hunk 16 above)
Comment (alxhub) [reply packages/compiler-cli/src/ngtsc/typecheck/src/type_check_block.ts]: Agreed. Renamed them all!

File: packages/compiler-cli/ngcc/src/analysis/decoration_analyzer.ts (hunk 18 above)
Comment (alxhub) [reply packages/compiler-cli/ngcc/src/analysis/decoration_analyzer.ts]: Fixed, thanks!

File: packages/compiler-cli/src/ngtsc/metadata/src/api.ts (hunk 19 above)
Comment (alxhub) [reply packages/compiler-cli/src/ngtsc/metadata/src/api.ts]: Updated the comment :)

File: packages/compiler-cli/src/ngtsc/metadata/src/util.ts (hunk 20 above)
Comment (alxhub) [reply packages/compiler-cli/src/ngtsc/metadata/src/util.ts]: I added the comment. The `CompoundMetadataReader` combines readers that represent multiple sources (e.g. the local registry, and .d.ts files). In practice it's likely only one of them will have data for a particular class.

File: packages/compiler-cli/test/ngtsc/template_typecheck_spec.ts (hunk 21 above)
Comment (alxhub) [reply packages/compiler-cli/test/ngtsc/template_typecheck_spec.ts]: I changed the selectors to make it clear :)

File: packages/compiler-cli/src/ngtsc/typecheck/src/expression.ts (hunk 22 above)
Comment (alxhub) [reply packages/compiler-cli/src/ngtsc/typecheck/src/expression.ts]: Not if `a` is a method call expression like `foo()` where `foo` can return `null`:

```typescript
(foo() != null ? foo().bar : undefined)
```

This will give an error because `foo()` doesn't have a narrowed type in the true branch of the ternary. And TypeScript is correct - just because the method returned non-null the first time it was called does not mean that it will the second time. The `!` is necessary, as a result.

I added a comment explaining it a bit.

**Comment by alxhub** [comment]
I cannot set the google3 status to successful:

> "This SHA and context has reached the maximum number of statuses."
